daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala

```diff
@@ -526,4 +526,18 @@ object ValueCoder {
   ): Either[DecodeError, Value] = {
     decodeValue(decodeCid, proto.VersionedValue.parseFrom(bytes))
   }
+
+  // Serialization of protobuf messages exceeding 2GB fails with different RuntimeExceptions.
+  // This functions tries to handle more gracefully those errors.
+  def toByteString(message: AbstractMessageLite[_, _]): Either[EncodeError, ByteString] =
+    try {
+      Right(message.toByteString)
+    } catch {
+      case _: RuntimeException =>
+        Left(
+          EncodeError(
+            s"Protobuf serialization of ${message.getClass.getName} failed: the message is probably too big."
```

Comment:
At least it won't throw anymore.

To make our support business easier, could you please:
- Embed the original exception (including stack trace) into the EncodeError. 
- Distinguish two cases: (1) message is too big (2) anything else. That way, we won't confuse our users with misleading assumptions of why an error occurred.

Reply:
>     * Embed the original exception (including stack trace) into the EncodeError.

I will add the original exception. 
On the other hand, I am a bit reluctant to add the stack trace, as it does not fit very well in our current design of errors. Let us think a bit more about that.

>     * Distinguish two cases: (1) message is too big (2) anything else. That way, we won't confuse our users with misleading assumptions of why an error occurred.

The problem is there no simple way to detect if the message is too big. 
Basically the Protobuf library does not handle overflow when computing the message size, so when the message is too big, the computed size is just meaningless. 

In practice, I was able to make the proto library crashes with two different errors : 
- `java.lang.NegativeArraySizeException: -2147483570"`
- `java.lang.RuntimeException: Serializing com.daml.lf.value.ValueOuterClass$Value to a ByteString threw an IOException (should never happen).`

I prefer not :
- I cannot guarantee those errors happen only in case of too big message,
- I cannot guarantee no other exception would be thrown in case of too big message,
- The only way I can distinguish the second exception from others `RuntimeException`, is by matching the error message, I think it is terrible idea.

Reply:
Difficult situation...

If protobuf throws an exception in production or at a client, I need the stack-trace to help the client work around it. Without a stack trace it gets pretty hard. I also understand that you don't want to copy the stack-trace into the error message. 

How about this:
- For the expected errors (message too big), you discard the stack-trace.
- For the unexpected errors (everything else), you won't catch the exception so it is up to the caller to handle it.

(At some point, it would probably good to inject a logger into this class and just log the exception details. But this is a major breaking change and therefore probably out-of-scope for the moment.)

Now about distinguishing "message too big" from other errors. How about this:
* NegativeArraySizeException gets mapped to "message too big".
* RuntimeException: If I recall it correctly, if this is due to "message too big", there should be an ArrayIndexOutOfBoundsException further down the cause chain. So if you find a RuntimeException with ArrayIndexOutOfBoundsException as a cause, you map it to "message too big". Otherwise, you map it to "everything else".

I agree that we don't have 100% certainty how to detect a "message too big" scenario. I still think we should do our best to detect this case, because then the user can resolve the situation without getting back to us.

Reply:
Matthias suggestion of discarding stack traces only on known exceptions makes sense to me. Throwing away stacktraces for exceptions that can occur from other stuff seems worse than letting them propagate to the caller.